core/src/bank_cleanup_service.rs

```diff
@@ -0,0 +1,81 @@
+// Service to purge 0 lamport accounts from the accounts_db.
+
+use crate::service::Service;
+use solana_ledger::bank_forks::BankForks;
+use solana_measure::measure::Measure;
+use std::string::ToString;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::{Arc, RwLock};
+use std::thread;
+use std::thread::{Builder, JoinHandle};
+use std::time::Duration;
+
+pub struct BankCleanupService {
+    t_cleanup: JoinHandle<()>,
+}
+
+impl BankCleanupService {
+    pub fn new(bank_forks: &Arc<RwLock<BankForks>>, exit: &Arc<AtomicBool>) -> Self {
+        info!("BankCleanupService active.");
+        let exit = exit.clone();
+        let bank_forks = bank_forks.clone();
+        let t_cleanup = Builder::new()
+            .name("solana-bank-cleanup".to_string())
+            .spawn(move || loop {
+                if exit.load(Ordering::Relaxed) {
+                    break;
+                }
+                let (bank, root) = {
+                    let bank_forks = bank_forks.read().unwrap();
+                    let banks = bank_forks.frozen_banks();
+                    let root = bank_forks.root();
+                    (banks[&root].clone(), root)
+                };
+                let mut time = Measure::start("purge_zero_lamport_accounts");
+                bank.purge_zero_lamport_accounts();
```

Comment:
Is the current accounts logic guaranteed to never successfully load a zero-balance account? I don't see a filter for zero balance accounts for instance in `load_slow` (might be missing something :D)

If we are guaranteed the above, then I think this is safe. If not, we might observe differing behavior based on when this cleanup runs.

Reply:
It gets filtered in accounts.rs

Reply:
https://github.com/solana-labs/solana/blob/master/runtime/src/accounts.rs#L270

Reply:
So this is called down to by `load_and_execute_transactions -> load_accounts` https://github.com/solana-labs/solana/blob/master/runtime/src/accounts.rs#L236. I think I'm missing where the filter happens in this path.

Reply:
I believe in that case it does an uwrap_or_default(), so a 0 account should end up having the same behavior as loading the tombstone account with 0 lamports.